### PR TITLE
sd-bus: rework sd_bus_error_set_errnofv()

### PR DESCRIPTION
**Pull request by yuwata**
This drops free() for constant string.

Replaces #16843.

### CODE REVIEW

**Comment by keszybz** [comment]
Hmm, could we please add test cases that test all the code paths in this function? The logic is quite complicated and we should have a unittest for it.

**Comment by yuwata** [comment]
@keszybz I noticed that `free(e->message)` is harmless, as `e->message` is always `NULL` at that point. So, I simply use @amitanandd's original commit, but its commit message is modified.

Also, tests are added.

PTAL.

File: src/libsystemd/sd-bus/bus-error.c [hunk 1]
```diff
@@ -536,7 +536,6 @@ _public_ int sd_bus_error_set_errnofv(sd_bus_error *e, int error, const char *fo
 
                                 free(m);
                         } else {
-                                free((char*) e->message);
```
Comment (poettering) [inline src/libsystemd/sd-bus/bus-error.c:539]: indeed!

File: src/libsystemd/sd-bus/bus-error.c [hunk 2]
```diff
@@ -515,33 +514,30 @@ _public_ int sd_bus_error_set_errnofv(sd_bus_error *e, int error, const char *fo
...
-                char *m;
+                _cleanup_free_ char *m = NULL;
 
                 /* Then, let's try to fill in the supplied message */
 
                 errno = error; /* Make sure that %m resolves to the specified error */
-                r = vasprintf(&m, format, ap);
-                if (r >= 0) {
+                if (vasprintf(&m, format, ap) < 0)
+                        goto failure;
```
Comment (poettering) [inline src/libsystemd/sd-bus/bus-error.c]: doesn't matter, but we more often than not, call that goto target just "fail", i.e without the "ure"

**Review by poettering — APPROVED** [review]

**Comment by yuwata** [comment]
@poettering Updated. The label is renamed: `failure` -> `fail`. Setting the green label.

**Comment by keszybz** [comment]
LGTM.

**Pull request merged**